torch/lib/ATen/DLConvertor.cpp

```diff
@@ -125,28 +125,41 @@ static ScalarType getATenScalarType(const DLDataType& dtype) {
 }
 
 
-// This function returns a shared_ptr to DLpack tensor constructed out ATen tensor
-DLTensor* toDLPack(const Tensor& src, DLTensor* dlTensor) {
-  dlTensor->data = src.data_ptr();
+void destructor(void * ctx, void * arg) {
+  static_cast<TensorImpl*>(ctx)->release();
+  delete static_cast<DLManagedTensor*>(arg);
+}
+
+
+// This function returns a shared_ptr to DLpack tensor constructed out of ATen tensor
+DLManagedTensor* toDLPack(const Tensor& src) {
+  // retain the underlying storage to make sure it doesn't go out of scope
```

Comment:
We should not be calling retain and release on at::Tensor directly. Instead we should just keep a struct that holds the tensor:

```
struct ATenDLTensor {
   at::Tensor handle; // this keeps the aten tensor alive
   DLManagedTensor tensor; // this is what gets returned
};
....
ATenDLTensor * dlMTensor = new ATenDLTensor;
dlMTensor.handle = src;
dlMTensor->tensor->ctx = dlMTensor;
dlMTensor->tensor->destructor = &destructor;
...
void destructor(void * ctx, void * arg) {
  delete static_cast<AtenDLTensor*>(ctx);
}
...
return &dlMTensor.tensor;
```

Reply:
just curious: why can't we call retain/release on at::Tensor directly? after all, we care about keeping the underlying storage is scope and retain/release can take care of refcounting as well?

Reply:
we discussed offline about this and having ATenDLTensor is a good idea because we don't want to expose the get(), retain(), release() members for ATen which are supposed to be private but marking them private right now is a pain because of friend structs.